### DISCUSSION TITLE
Clarification needed about using the SPI hardware protocol

### DISCUSSION THREAD

**Discussion by nickthiru** [Moddable-OpenSource/moddable#1059] [Q&A] [answered]
Hi,

I am in the process of reviewing issues https://github.com/Moddable-OpenSource/moddable/discussions/989 and https://github.com/Moddable-OpenSource/moddable/discussions/991 so that I can implement the `sx127x` driver with an `esp32/nodemcu`.

I believe it is possible to use the SPI interface with an SX1276 LoRa module but then according to https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/pins/pins.md#spi, there is no SPI API. The lora [example](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/drivers/lora) and the [provider](https://github.com/Moddable-OpenSource/moddable/blob/public/build/devices/esp32/targets/nodemcu/host/provider.js#L29) for the `nodemcu/host` too seem to be set to work with SPI as well.

Am I missing something, in terms of the difference between an SPI API and being able to set the SPI pins and use them?

**Comment by andycarle** [comment, marked as answer]
Hi @nickthiru,

The comment you surfaced from the Pins documentation is only relevant to our older API for working with hardware pins. Most of the new work going into Moddable SDK projects and examples is instead based on the newer ECMA-419 standard and the Moddable SDK implementation of that standard.

SPI works well with the ECMA-419 I/O modules, as you've seen from the code you've discovered. There are also examples available [here](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/io/spi). 

The Documentation for ECMA-419 SPI is just the spec itself. [This is the relevant section.](https://419.ecma-international.org/#-10-io-classes-serial-peripheral-interface-spi)

**Reply by nickthiru** [reply]
@andycarle Thanks! Wow, I completely missed this, and look forward to diving into it.

As a sidenote, is there anything on using the UART protocol? I have a NEO-6M GPS tracker that I was told uses UART, which I have not used before. I did find [this](https://ecmatc53.github.io/spec/web/spec.html#-10-io-classes-serial), which is reflected in the [io.md](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/io/io.md#serial) doc as well, but I'm wondering how to use it. Can I just `import Serial from "embedded:io/serial";` and use the API?

**Reply by nickthiru** [reply]
@andycarle Please ignore the second paragraph from the previous post. I am studying `io.md` and will try and figure it out myself first. Thank you.

**Reply by phoddie** [reply]
@nickthiru – Just FYI, the PZEM0004T sensor implementation (by @andycarle!) is a good example of using ECMA-419 serial to communicate with a sensor. Check out both the [driver](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/drivers/sensors/pzem-004t/pzem-004t.js) and the [example app](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/drivers/sensors/pzem-004t/console/main.js).

**Reply by nickthiru** [reply]
@phoddie Great stuff! Thank you very much.
